### PR TITLE
Change the packaging to jar and configure a war profile

### PR DESCRIPTION
**Pull request by PierreBesson**
This needs review and a lot of testing to make sure it works for every case.

In order to build the war, you can run:
`./mvnw -Pprod,war verify`

In my testing, running the produced war file in tomcat and deploying to GAE worked well.

-   [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [ ] Tests are added where necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

**Comment by PierreBesson** [comment]
For gradle I have had a look at it but I'm not sure how to do it : @atomfrede do you have an idea ?

**Comment by atomfrede** [comment]
@PierreBesson Just to make sure, the title of this PR is wrong, right? You mean `Change the packaging to **jar** for maven and configure a war profile`?
For gradle would not use a `-P` switch. Need to check but I think we can just change the default packing to jar. As we still apply the `war` plugin there is `bootWar` task available. Everything that needs a war file must just call bootWar instead bootJar or assemble, but will have look to check if thats correct.

**Comment by chanseokoh** [comment]
The title says "Change the packaging to war" and I was confused until I actually see the code. It should say "Change the packaging to jar".

**Comment by atomfrede** [comment]
@loosebazooka We currently have `war` and `jar` packaging available, just a matter of calling `bootJar` directly. I would prefer not to add more hidden flags as they are not discoverable via `gradle tasks`, but maybe that needs a vote.

**Comment by chanseokoh** [comment]
@atomfrede @PierreBesson is it possible to remove the WAR Plugin for Gradle?

```diff
 apply plugin: 'maven'
 apply plugin: 'org.springframework.boot'
-apply plugin: 'war'
 apply plugin: 'propdeps'
 apply plugin: 'com.moowork.node'
 apply plugin: 'io.spring.dependency-management'
```

**Comment by chanseokoh** [comment]
Actually, more places with WAR:

```diff
diff --git a/build.gradle b/build.gradle
index 6e4c913..798935d 100644
--- a/build.gradle
+++ b/build.gradle
@@ -34,7 +34,6 @@ assert System.properties['java.specification.version'] == '1.8'
 
 apply plugin: 'maven'
 apply plugin: 'org.springframework.boot'
-apply plugin: 'war'
 apply plugin: 'propdeps'
 apply plugin: 'com.moowork.node'
 apply plugin: 'io.spring.dependency-management'
@@ -54,14 +53,6 @@ version = '0.0.1-SNAPSHOT'
 
 description = ''
 
-bootWar {
-   mainClassName = 'com.example.demoapp.JhipsterApp'
-}
-
-war {
-
-}
-
 springBoot {
     mainClassName = 'com.example.demoapp.JhipsterApp'
     buildInfo()
@@ -255,9 +246,6 @@ wrapper {
     gradleVersion = '4.10.2'
 }
 
-task stage(dependsOn: 'bootWar') {
-}
-
 if (project.hasProperty('nodeInstall')) {
     node {
         version = "${node_version}"
diff --git a/gradle/profile_dev.gradle b/gradle/profile_dev.gradle
index a1b1eb5..22bbce8 100644
--- a/gradle/profile_dev.gradle
+++ b/gradle/profile_dev.gradle
@@ -28,10 +28,6 @@ task webpackBuildDev(type: NpmTask, dependsOn: 'npm_install') {
     args = ["run", "webpack:build"]
 }
 
-war {
-    webAppDirName = 'build/www/'
-}
-
 task copyIntoStatic (type: Copy) {
     from 'build/www/'
     into 'build/resources/main/static'
diff --git a/gradle/profile_prod.gradle b/gradle/profile_prod.gradle
index 26a2978..1eec73c 100644
--- a/gradle/profile_prod.gradle
+++ b/gradle/profile_prod.gradle
@@ -27,10 +27,6 @@ task webpack(type: NpmTask, dependsOn: 'npm_install') {
     args = ["run", "webpack:prod"]
 }
 
-war {
-    webAppDirName = 'build/www/'
-}
-
 task copyIntoStatic (type: Copy) {
     from 'build/www/'
     into 'build/resources/main/static'
```

**Comment by PierreBesson** [comment]
@chanseokoh will correct me if I'm wrong but I think the problem is as long as there is the war plugin enabled in build.gradle, jib (starting with v0.10) will assume that we want war packaging.

In my tests I have been able to work around that by removing all the war code highlighted above and added this to the build.gradle : 
```
if (project.hasProperty('war')) {
    apply plugin: 'war'
    war {
        enabled = true
    }
}
```
In this regard it will work similarly to maven. Should I go forward with this change ?

**Comment by chanseokoh** [comment]
> as long as there is the war plugin enabled in build.gradle, jib (starting with v0.10) will assume that we want war packaging.

That is correct.

> Should I go forward with this change ?

From the perspective of the Jib developer, that (not applying the WAR Plugin) is the way we want.

Just from my personal curiosity, is `war.enabled = true` necessary?

**Comment by atomfrede** [comment]
@PierreBesson I would like to remove most `-P` flags, but in that case it seems we need to have it.

**Comment by PierreBesson** [comment]
@chanseokoh No it is actually not required. I'm not sure why it was there but it could very well be because it was needed for older Spring Boot versions. JHipster actually started with very early versions of Spring Boot.

**Comment by atomfrede** [comment]
@chanseokoh No I think I didn't remove it. According to the boot documentation this is supposed to enable war creation in addition to jar creation, but I realized is has no effect (maybe because the war plugin is already applied). If jib can handle executable jars and we do not support container deploymend anymore we can definitl;y remove the war plugin as there are bootJar and bootWar available via the boot plugin.

**Comment by PierreBesson** [comment]
@atomfrede That is not correct, if you don't add the `apply plugin: war` then the spring boot plugin won't show the bootWar task so I think the solution using -Pwar is required.

**Comment by atomfrede** [comment]
+1 was not sure to be honest.

File: generators/common/templates/README.md.ejs [hunk 1]
```diff
@@ -293,7 +302,7 @@ To stop it and remove the container, run:
 You can also fully dockerize your application and all the services that it depends on.
 To achieve this, first build a docker image of your app by running:
 
-    <% if (buildTool === 'maven') { %>./mvnw package -Pprod verify jib:dockerBuild<% } %><% if (buildTool === 'gradle') { %>./gradlew bootWar -Pprod jibDockerBuild<% } %>
+    <% if (buildTool === 'maven') { %>./mvnw package -Pprod verify jib:dockerBuild<% } %><% if (buildTool === 'gradle') { %>./gradlew bootJar -Pprod jibDockerBuild<% } %>
```
Comment (chanseokoh) [inline generators/common/templates/README.md.ejs]: FYI: `bootJar` may not be necessary. Jib doesn't look into a packaged JAR, so normally it's enough to do `./gradlew jibDockerBuild`, unless you wired up extra tasks with `bootJar`.

File: generators/docker-base.js [hunk 2]
```diff
@@ -49,7 +49,7 @@ function checkImages() {
             runCommand = './mvnw package -Pprod verify jib:dockerBuild';
         } else {
             imagePath = this.destinationPath(`${this.directoryPath + appsFolder}/build/jib-cache`);
-            runCommand = './gradlew bootWar -Pprod jibDockerBuild';
+            runCommand = './gradlew bootJar -Pprod jibDockerBuild';
```
Comment (chanseokoh) [inline generators/docker-base.js]: ditto

File: generators/docker-base.js [hunk 3]
```diff
@@ -49,7 +49,7 @@ function checkImages() {
             runCommand = './mvnw package -Pprod verify jib:dockerBuild';
```
Comment (chanseokoh) [inline generators/docker-base.js]: Related, `package` is no longer required for the Jib standard (non-WAR) containerization. Normally, `./mvnw compile jib:dockerBuild` should work (although if we were to do the WAR containerization, `package` is required.)

File: generators/common/templates/README.md.ejs (hunk 1 above)
Comment (PierreBesson) [reply generators/common/templates/README.md.ejs]: I wished our build could be as fast as just compiling the sources but we have a complex builds with resource copying and filtering and in some case even a front-end nodejs build to run. However I wish we could create a special mode to speed up everything and have a fast dev workflow with jib + scaffold. Anyway thanks a lot for your help !

**Comment by pascalgrimaud** [comment]
Good job @PierreBesson !
It is an important PR, but should I start to test everything now, or should I wait for the merge of Spring Boot 2.1 branch ?

**Comment by PierreBesson** [comment]
This is not related to Spring boot 2.1 but maybe it might be better to wait for it to be merged as this will need extensive manual testing.
The things to test are :
- [ ] running the app as jar
- [ ] running the app as war in an app server
- [ ] running the app as docker image
- [ ] deploy to heroku, gae, etc..
- [ ] ci-cd in Jenkins, gitlab...

**Comment by pascalgrimaud** [comment]
I'll start to review this, but as there are some conflicts, can you rebase your branch plz @PierreBesson ?
Tell me when it's done

**Comment by PierreBesson** [comment]
The rebase is done.

**Comment by pascalgrimaud** [comment]
Tested and it works well.
I'm going to merge this, waiting the CI to finish

**Comment by pascalgrimaud** [comment]
good job @PierreBesson !

**Comment by DanielFran** [comment]
@PierreBesson @pascalgrimaud @jdubois Did not found any update in documentation about those changes...

**Comment by pascalgrimaud** [comment]
@DanielFran : maybe we forgot to do it... Can you open a new ticket plz ? So we won't forget to do it

**Comment by DanielFran** [comment]
@pascalgrimaud Done!

**Pull request merged**